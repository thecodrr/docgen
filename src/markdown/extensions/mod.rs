#[macro_use]
pub mod macros;
pub mod callout;
pub mod codeblock;
pub mod emoji;
pub mod link_rewriter;
pub mod math;
pub mod mermaid;
pub mod tabs;
pub mod task_list;
pub mod toc;
