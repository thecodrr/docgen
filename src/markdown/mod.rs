mod extension;
pub mod extensions;
pub mod parser;
