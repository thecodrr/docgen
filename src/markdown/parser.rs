use std::collections::HashMap;

use pulldown_cmark::{html, CowStr, Event, Options, Parser};

use super::{
    extension::{Extension, Output, TextExtension},
    extensions::{
        callout::Callout,
        codeblock::CodeBlock,
        emoji::EmojiConverter,
        link_rewriter::{Link, LinkRewriter},
        math::MathBlock,
        mermaid::MermaidBlock,
        tabs::Tabs,
        toc::{Heading, TableOfContents},
    },
};

pub struct MarkdownParser {
    pub extensions: Vec<Box<dyn Extension>>,
    pub text_processors: Vec<Box<dyn TextExtension>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ParsedMarkdown {
    pub html: String,
    pub headings: Vec<Heading>,
    pub links: Vec<Link>,
}

pub struct ParseOptions {
    /// Changes the root URL for any links that point to the current domain.
    pub url_root: String,
    pub link_rewrite_rules: HashMap<String, String>,
    pub url_params: Vec<(String, String)>,
    pub root_dir: Option<String>,
    // pub resolve_embeds: Option<Box<dyn Fn(String) -> Option<String>>>,
}

impl Default for ParseOptions {
    fn default() -> Self {
        ParseOptions {
            url_root: String::from("/"),
            link_rewrite_rules: HashMap::new(),
            url_params: vec![],
            root_dir: None,
        }
    }
}

impl MarkdownParser {
    pub fn new(options: Option<ParseOptions>) -> Self {
        let parse_opts = options.unwrap_or(ParseOptions::default());

        let url_root = parse_opts.url_root.to_owned();
        let link_rewrite_rules = parse_opts.link_rewrite_rules.to_owned();
        let url_params = parse_opts.url_params.to_owned();

        let extensions: Vec<Box<dyn Extension>> = vec![
            Box::new(Callout),
            Box::new(MermaidBlock),
            Box::new(MathBlock),
            Box::new(Tabs {
                current_tabgroup: None,
                current_tab: None,
            }),
            Box::new(CodeBlock),
            Box::new(LinkRewriter {
                url_root,
                link_rewrite_rules,
                url_params,
                current_link: None,
            }),
            Box::new(TableOfContents {
                current_heading: None,
            }),
        ];

        let text_processors: Vec<Box<dyn TextExtension>> = vec![Box::new(EmojiConverter)];

        MarkdownParser {
            extensions,
            text_processors,
        }
    }

    pub fn parse(&mut self, input: &str) -> ParsedMarkdown {
        let mut parser = Parser::new_ext(input, Options::all())
            .into_iter()
            .peekable();

        let mut events: Vec<Event> = Vec::new();
        let mut links: Vec<Link> = Vec::new();
        let mut headings: Vec<Heading> = Vec::new();

        while let Some(ev) = parser.next() {
            let event = &mut ev.to_owned();

            match event {
                Event::Text(text) => {
                    let mut copy = text.to_string();
                    for extension in &self.text_processors {
                        copy = extension.process_text(&copy);
                        *event = Event::Text(CowStr::from(copy.to_owned()));
                    }
                }
                Event::Html(_) => {
                    continue;
                }
                _ => {}
            }

            let mut handled = false;
            for extension in &mut self.extensions {
                let (output, is_handled) = extension.process_event(&mut events, &event);

                handle_output(output, &mut events, &mut links, &mut headings);

                if is_handled {
                    handled = true;
                    break;
                }
            }

            if !handled {
                events.push(event.to_owned());
            }
        }

        for extension in &mut self.extensions {
            let output = extension.end_of_doc(&mut events);
            handle_output(output, &mut events, &mut links, &mut headings);
        }

        // Write to String buffer.
        let mut as_html = String::new();
        html::push_html(&mut as_html, events.into_iter());

        ParsedMarkdown {
            html: as_html,
            headings,
            links,
        }
    }
}

fn handle_output<'a>(
    output: Option<Vec<Output<'a>>>,
    events: &mut Vec<Event<'a>>,
    links: &mut Vec<Link>,
    headings: &mut Vec<Heading>,
) {
    if let Some(output) = output {
        output.into_iter().for_each(|result| match result {
            Output::Event(ev) => events.push(ev),
            Output::Link(link) => links.push(link),
            Output::Heading(heading) => headings.push(heading),
            _ => {}
        });
    }
}
