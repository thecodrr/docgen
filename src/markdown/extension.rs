use pulldown_cmark::{CowStr, Event};

use super::extensions::{link_rewriter::Link, toc::Heading};

pub enum Output<'a> {
    None,
    Event(Event<'a>),
    Link(Link),
    Heading(Heading),

    Block(&'a str),
}

pub trait Extension {
    fn process_event<'a>(
        &mut self,
        events: &mut Vec<Event<'a>>,
        event: &Event<'a>,
    ) -> (Option<Vec<Output<'a>>>, bool);

    fn end_of_doc<'a>(&mut self, _events: &mut Vec<Event<'a>>) -> Option<Vec<Output<'a>>> {
        None
    }
}

pub trait TextExtension {
    fn process_text(&self, text: &CowStr) -> CowStr;
}
