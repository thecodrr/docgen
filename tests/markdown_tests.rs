use docgen::markdown::parser::{MarkdownParser, ParseOptions};
use insta::*;

#[macro_use]
extern crate indoc;

#[cfg(test)]
mod test {
    use super::*;

    macro_rules! snapshot_test {
        ($name:ident, $input:expr, $options:expr) => {
            #[test]
            fn $name() {
                let input = indoc! {$input};
                let mut options = ParseOptions::default();
                ($options)(&mut options);

                insta::with_settings!({
                    description => stringify!($name),
                    info => &input,
                    omit_expression => true // do not include the default expression
                }, {

                    let mut parser = MarkdownParser::new(Some(options));
                    assert_debug_snapshot!(parser.parse(&input));
                });
            }
        };
    }

    snapshot_test!(
        supports_headings,
        "# My heading

    Some content

    ## Some other heading",
        |_| {}
    );

    snapshot_test!(supports_links, "\n[an link](/foo/bar)\n", |_| {});

    snapshot_test!(
        rewrite_link_root_path,
        "\n[an link](/foo/bar)\n",
        |options: &mut ParseOptions| {
            options.url_root = "/other/root".to_owned();
        }
    );

    snapshot_test!(
        does_not_rewrite_non_absolute_urls,
        "\n[an link](https://www.google.com)\n",
        |options: &mut ParseOptions| {
            options.url_root = "/other/root".to_owned();
        }
    );

    snapshot_test!(
        does_not_rewrite_relative_urls,
        "\n[a relative link](relative/link)\n",
        |options: &mut ParseOptions| {
            options.url_root = "/other/root".to_owned();
        }
    );

    snapshot_test!(
        rewrites_any_image_that_has_an_explicit_rewrite_mapping,
        "\n![an image](/assets/cat.jpg)\n",
        |options: &mut ParseOptions| {
            options.link_rewrite_rules.insert(
                "/assets/cat.jpg".to_owned(),
                "https://example.com/cat.jpg".to_owned(),
            );
        }
    );

    snapshot_test!(
        rewrites_any_link_that_has_an_explicit_rewrite_mapping,
        "\n[an document](/assets/plans.pdf)\n",
        |options: &mut ParseOptions| {
            options.link_rewrite_rules.insert(
                "/assets/plans.pdf".to_owned(),
                "https://example.com/plans.pdf".to_owned(),
            );
        }
    );

    snapshot_test!(
        appends_parameters_to_the_end_of_urls,
        "\n[an link](relative/link)\n",
        |options: &mut ParseOptions| {
            options
                .url_params
                .push(("base".to_owned(), "123".to_owned()));
        }
    );

    snapshot_test!(
        appends_multiple_parameters_to_the_end_of_urls,
        "\n[an link](relative/link)\n",
        |options: &mut ParseOptions| {
            options
                .url_params
                .push(("bases".to_owned(), "are".to_owned()));
            options
                .url_params
                .push(("belong".to_owned(), "tous".to_owned()));
        }
    );

    snapshot_test!(
        appends_multiple_parameters_to_the_end_of_absolute_urls,
        "\n[an link](/absolute/link)\n",
        |options: &mut ParseOptions| {
            options
                .url_params
                .push(("bases".to_owned(), "are".to_owned()));
            options
                .url_params
                .push(("belong".to_owned(), "tous".to_owned()));
        }
    );

    snapshot_test!(
        does_not_append_params_to_urls_with_a_specific_domain,
        "\n[an link](http://www.example.com/)\n",
        |options: &mut ParseOptions| {
            options
                .url_params
                .push(("bases".to_owned(), "are".to_owned()));
            options
                .url_params
                .push(("belong".to_owned(), "tous".to_owned()));
        }
    );

    snapshot_test!(
        sanitizes_input,
        "<script>
        alert('I break you');
        </script>",
        |_| {}
    );

    snapshot_test!(
        allows_mermaid_blocks,
        "```mermaid
        graph TD;
            A-->B;
            A-->C;
        ```",
        |_| {}
    );

    snapshot_test!(
        allows_code_blocks,
        "```ruby
        1 + 1
        something else
        something else too
        another something else
        ```",
        |_| {}
    );

    snapshot_test!(
        gathers_a_list_of_links_on_the_page,
        "[foo](/bar)
        [Example](https://www.example.com)
        [Example 2](https://www.example2.com)",
        |_| {}
    );

    snapshot_test!(
        gathers_the_internal_text_of_a_link,
        "[**BOLD**](/bar)
        [![AltText](/src/foo)](/bar)
        ## [AnHeader](/bar)",
        |_| {}
    );

    snapshot_test!(detects_emojis, "I am :grinning:.", |_| {});

    snapshot_test!(detects_emojis_in_links, "[:grinning:](/foo)", |_| {});

    snapshot_test!(
        leaves_the_emoji_identifier_alone_if_it_is_not_recognised,
        "Look at this :idonotexist:",
        |_| {}
    );

    snapshot_test!(
        ignores_identifiers_that_do_not_end,
        "Look at this :stop",
        |_| {}
    );

    snapshot_test!(
        ignores_identifiers_that_do_not_end_with_whitespace,
        "Look at this :stop MORE",
        |_| {}
    );

    snapshot_test!(
        supports_tabgroups,
        "# [_Tab1_](#/tab/id1/condition1)
        Foo
        # [Tab2](#/tab/id2)
        Bar

        ---",
        |_| {}
    );

    snapshot_test!(
        supports_multiple_tabgroups,
        "# [_Tab1_](#/tab/id1/condition1)
        Foo
        # [Tab2](#/tab/id2)
        Bar

        ---
        
        # [_Tab1_](#/tab/id1/condition1)
        Foo
        # [Tab2](#/tab/id2)
        Bar

        ---
        
        After content",
        |_| {}
    );

    snapshot_test!(
        supports_content_around_tabgroups,
        "content before

        # [_Tab1_](#/tab/id1/condition1)
        Foo
        # [Tab2](#/tab/id2)
        Bar

        ---
        
        content after",
        |_| {}
    );

    snapshot_test!(
        last_tabgroup_in_doc_does_not_require_ending_rule,
        "content before

        # [_Tab1_](#/tab/id1/condition1)
        Foo
        # [Tab2](#/tab/id2)
        Bar",
        |_| {}
    );

    snapshot_test!(
        supports_headings_inside_tabgroups,
        "content before

        # [_Tab1_](#/tab/id1/condition1)
        
        ## Heading inside tab 1

        some content

        # [Tab2](#/tab/id2)
        Bar

        ## Heading inside tab 2

        ---
        
        content after",
        |_| {}
    );

    snapshot_test!(
        callout_title_is_optional,
        "> warning
        > 
        > The content",
        |_| {}
    );

    snapshot_test!(
        callouts_can_have_stuff_after_it,
        "> warning An Note
        > 
        > The content

        Moar
        ",
        |_| {}
    );

    snapshot_test!(
        callouts_can_contain_images,
        "> info An Info
        >
        > ![an pic](/cat.jpg)",
        |_| {}
    );

    snapshot_test!(
        blockquotes_are_not_confused_with_callouts,
        "> I am here!",
        |_| {}
    );

    snapshot_test!(
        supports_github_style_markdown_checkboxes,
        "
        * [ ] Incomplete
        * [x] Complete
        ",
        |_| {}
    );

    snapshot_test!(
        does_not_allow_random_forms,
        "<form>
          <label for=\"ufname\">First name:</label><br>
          <input type=\"text\" id=\"fname\" name=\"fname\"><br>
          <label for=\"lname\">Last name:</label><br>
          <input type=\"text\" id=\"lname\" name=\"lname\">
        </form>",
        |_| {}
    );

    snapshot_test!(
        it_detects_math_blocks,
        "```math
        x^2 - 5x + 6 = 0 \\
        (x-2)(x-3)=0 \\
        \\textrm{then either }x=2 \\,or\\,x=3
        ```
        ",
        |_| {}
    );

    snapshot_test!(
        inline_code_in_headings_is_included_in_heading_titles,
        "# Foo `bar` baz
        ",
        |_| {}
    );

    // snapshot_test!(
    //     supports_markdown_source_embeds,
    // "I was working but I couldn't.

    // # Heading in main file

    // ![](/embed/file.md)

    // And some more content here.",
    //     |options: &mut ParseOptions| {
    // options.resolve_embeds = Some(Box::new(|url: String| match url.as_str() {
    //     "/embed/file.md" => Some("# Heading from another file".to_string()),
    //     _ => None,
    // }));
    //     }
    // );

    // #[test]
    // fn supports_markdown_source_embeds() {
    //     let input = indoc!(
    //         "I was working but I couldn't.

    //     # Heading in main file

    //     [embed ./html/hello.md](./html/hello.md)

    //     And some more content here."
    //     );

    //     insta::with_settings!({
    //         description => "Supports callout blocks",
    //         info => &input,
    //         omit_expression => true // do not include the default expression
    //     }, {
    //         let mut parser = MarkdownParser::new(None);
    //         assert_debug_snapshot!(parser.parse(&input));
    //     });
    // }

    #[test]
    fn supports_callout_blocks() {
        let binding = ["info", "notice", "success", "warn", "warning", "error"]
            .map(|kind| {
                formatdoc! {"> {} An Note
                >
                > The content
                >
                > More content", kind}
            })
            .join("\n\n---\n\n");

        let input = binding.as_str();

        insta::with_settings!({
            description => "Supports callout blocks",
            info => &input,
            omit_expression => true // do not include the default expression
        }, {
            let mut parser = MarkdownParser::new(None);
            assert_debug_snapshot!(parser.parse(&input));
        });
    }
}
