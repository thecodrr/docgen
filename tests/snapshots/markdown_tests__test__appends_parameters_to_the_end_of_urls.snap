---
source: tests/markdown_tests.rs
description: appends_parameters_to_the_end_of_urls
info: "\n[an link](relative/link)\n"
---
ParsedMarkdown {
    html: "<p><a href=\"relative/link?base=123\">an link</a></p>\n",
    preview: "an link",
    headings: [],
    links: [
        Link {
            title: "an link",
            url: Local(
                "relative/link?base=123",
            ),
        },
    ],
    blocks: {},
}
